likelihood.py

```diff
@@ -0,0 +1,232 @@
+            line = datafile.readline()
+            while line.find('#') != -1:
+                line = datafile.readline()
+            self.W0[i] = float(line.split()[0])
+            self.W2[i] = float(line.split()[1])
+            self.W4[i] = float(line.split()[2])
+        datafile.close()
+
+        # Precompute useful window function things
+        kmax = 100.
+        self.k0 = 5.e-4
+
+        self.rmin = 0.01
+        rmax = 1000.
+        b = -1.1001
+        bR = -2.001
+
+        Delta = log(kmax/self.k0) / (self.Nmax - 1)
+        Delta_r = log(rmax/self.rmin) / (self.Nmax - 1)
+        i_arr = np.arange(self.Nmax)
+        rtab = self.rmin * exp(Delta_r * i_arr)
+
+        self.kbins3 = self.k0 * exp(Delta * i_arr)
+        self.tmp_factor = exp(-1.*b*i_arr*Delta)
+        self.tmp_factor2 = exp(-1.*bR*i_arr*Delta_r)[:,np.newaxis]
+
+        jsNm = np.arange(-self.Nmax//2,self.Nmax//2+1,1)
+        self.etam = b + 2*1j*pi*(jsNm)/self.Nmax/Delta
+
+        def J_func(r,nu):
+            gam = special.gamma(2+nu)
+            r_pow = r**(-3.-1.*nu)
+            sin_nu = np.sin(pi*nu/2.)
+            J0 = -1.*sin_nu*r_pow*gam/(2.*pi**2.)
+            return J0
+
+        j0 = J_func(rtab.reshape(-1,1),self.etam.reshape(1,-1))
+        self.J0_arr = j0[:,:,np.newaxis]
+
+        self.etamR = bR + 2*1j*pi*(jsNm)/self.Nmax/Delta_r
+
+        def Jk_func(k,nu):
+            gam = special.gamma(2+nu)
+            k_pow = k**(-3.-1.*nu)
+            sin_nu = np.sin(pi*nu/2.)
+            J0k = -1.*k_pow*gam*sin_nu*(4.*pi)
+            return J0k
+
+        j0k = Jk_func(self.kbins3.reshape(-1,1),self.etamR.reshape(1,-1))
+        self.J0k_arr = j0k[:,:,np.newaxis]
+
+        #Precompute inverse covariance matrix
+        self.invcov = np.linalg.inv(self.cov)
+
+        # Compute window response matrix
+        self.response_matrix = np.zeros((self.ksize,self.Nmax))
+        for i in range(self.Nmax):
+            tmp_resp0 = self.window_response(i)
+            self.response_matrix[:,i] = tmp_resp0[0]
+
+        # Now convolve with window function
+        self.invcovW = np.matmul(self.response_matrix.T,self.invcov)
+        self.invcovWW = np.matmul(self.response_matrix.T,np.matmul(self.invcov,self.response_matrix))
+
+    def window_response(self,k_index):
+        """Window function. This is from 1607.03150 and includes only the monopole.
+        Accurate for k > 0.0015."""
+        Nmax = self.Nmax
+        k0 = self.k0
+
+        Pdiscrin0 = np.zeros(Nmax)
+        Pdiscrin0[k_index] = 1
+
+        cm0 = np.fft.fft(Pdiscrin0)/ Nmax
+        cmsym0 = np.zeros(Nmax+1,dtype=np.complex_)
+
+        all_i = np.arange(Nmax+1)
+        f = (all_i+2-Nmax//2) < 1
+        cmsym0[f] = k0**(-self.etam[f])*np.conjugate(cm0[-all_i[f]+Nmax//2])
+        cmsym0[~f] = k0**(-self.etam[~f])*cm0[all_i[~f]-Nmax//2]
+
+        cmsym0[-1] = cmsym0[-1] / 2
+        cmsym0[0] = cmsym0[0] / 2
+
+        xi0 = np.real(cmsym0*self.J0_arr).sum(axis=1)
```

Comment:
Suggest:
```python
xi0 = np.real(cmsym0[:, None]*self.J0_arr).sum(axis=1)
```

`cmsym0` is (129, ); `self.J0_arr` is (128, 129, 1).
If you don't add a new dimension for `cmsym0`, 
`cmsym0*self.J0_arr` will be (128, 129, 129).
But I think you want `xi0` to be (128, 1).